src/include/api/flat_l2_index.h

```diff
@@ -0,0 +1,296 @@
+/**
+ * * @file   api/flat_l2_index.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the `IndexFlatL2` class, which is a type-erased
+ * wrapper of `index_flat_l2` that allows for runtime polymorphism of the
+ * `index_flat_l2` class template.
+ *
+ * See README.md for details on type erasure.
+ */
+
+#ifndef TILEDB_API_FLAT_L2_INDEX_H
+#define TILEDB_API_FLAT_L2_INDEX_H
+
+#include "api/feature_vector.h"
+#include "api/feature_vector_array.h"
+#include "api_defs.h"
+#include "detail/linalg/tdb_helpers.h"
+#include "index/flat_l2_index.h"
+
+/**
+ * A type-erased index class for C++ ivf_flat_l2_index.
+ */
+class IndexFlatL2 {
```

Comment:
Shouldn't this be just `IndexFlat`? Why do we add the distance name in the class name when we want these classes to be using configurable distance functions?

Reply:
The C++ `ivf_index` class should be parameterized (will be parameterized) by distance function, but the type-erased class can't be type-parameterized in the same way a class template can.  I think Faiss has `indexFlatL2`, for instance (https://faiss.ai/cpp_api/struct/structfaiss_1_1IndexFlatL2.html). 

I created a story for this: https://app.shortcut.com/tiledb-inc/story/40228/parameterize-c-index-classes-and-their-type-erased-wrappers-on-distance-function